File: src/intersection.cpp
#include "intersection.h"
#include "form.h"
#include "ray.h"
#include "tuple.h"
#include "pch.h"
#include <algorithm>

IntersectionState::IntersectionState() {
	mbrTime = 0;
	mbrObject = std::make_unique<Form>();
	mbrPoint = Point();
	mbrEye = Vector();
	mbrNormal = Vector();
	mbrInside = false;
	mbrOverPoint = Point();
}
IntersectionState::IntersectionState(const IntersectionState &argOther) {
	mbrTime = argOther.mbrTime;
	setObject(argOther.mbrObject.get());
	mbrPoint = argOther.mbrPoint;
	mbrEye = argOther.mbrEye;
	mbrNormal = argOther.mbrNormal;
	mbrInside = argOther.mbrInside;
	mbrOverPoint = argOther.mbrOverPoint;
}
IntersectionState& IntersectionState::operator=(const IntersectionState &argOther) {
	if (this == &argOther) return *this;
	mbrTime = argOther.mbrTime;
	setObject(argOther.mbrObject.get());
	mbrPoint = argOther.mbrPoint;
	mbrEye = argOther.mbrEye;
	mbrNormal = argOther.mbrNormal;
	mbrInside = argOther.mbrInside;
	mbrOverPoint = argOther.mbrOverPoint;
	return *this;
}
void IntersectionState::setObject(Form* argObject) {
    if (Sphere *varSphere = dynamic_cast<Sphere *>(argObject))
    {
        mbrObject = std::make_unique<Sphere>(*varSphere);
    }
    else if (Plane *varPlane = dynamic_cast<Plane *>(argObject))
    {
        mbrObject = std::make_unique<Plane>(*varPlane);
    }
    else {
        mbrObject = std::make_unique<Form>(*argObject);
    }
}

Intersection::Intersection()
{
	mbrObject = std::make_unique<Form>();
	mbrTime = 0;
	mbrExists = true;
}
Intersection::Intersection(double time, Form *s)
{
	setObject(s);
	this->mbrTime = time;
	mbrExists = true;
}
// Intersection::Intersection(double time, std::unique_ptr<Form> s)
// {
// 	setObject(s.get());
// 	this->mbrTime = time;
// 	mbrExists = true;
// }
Intersection::Intersection(const Intersection &argOther) {
	setObject(argOther.mbrObject.get());
	mbrTime = argOther.mbrTime;
	mbrExists = argOther.mbrExists;
}
Intersection& Intersection::operator=(const Intersection argOther){
	if (this == &argOther) return *this;
	setObject(argOther.mbrObject.get());
	mbrTime = argOther.mbrTime;
	mbrExists = argOther.mbrExists;
	return *this;
}
bool Intersection::checkEqual(Intersection other)
{
	return mbrTime == other.mbrTime && mbrObject->checkEqual(*other.mbrObject.get());
}
IntersectionState Intersection::getState(Ray argRay)
{
	IntersectionState is = IntersectionState();
	is.mbrTime = mbrTime;
	is.setObject(mbrObject.get());
	is.mbrPoint = argRay.getPosition(is.mbrTime);
	is.mbrEye = -(argRay.mbrDirection);
	is.mbrNormal = is.mbrObject->getNormal(is.mbrPoint);
	if (is.mbrNormal.dot(is.mbrEye) < 0) {
		is.mbrInside = true;
		is.mbrNormal = -is.mbrNormal;
	}
	else {
		is.mbrInside = false;
	}
	is.mbrOverPoint = is.mbrPoint + is.mbrNormal * getEpsilon();
	return is;
}
void Intersection::setObject(Form* argObject) {
    if (Sphere *varSphere = dynamic_cast<Sphere *>(argObject))
    {
        mbrObject = std::make_unique<Sphere>(*varSphere);
    }
    else if (Plane *varPlane = dynamic_cast<Plane *>(argObject))
    {
        mbrObject = std::make_unique<Plane>(*varPlane);
    }
    else {
        mbrObject = std::make_unique<Form>(*argObject);
    }
}

Intersections::Intersections()
{
	//Intersection i = Intersection(0, Sphere());
	//this->intersections.push_back(i);
}
Intersections::Intersections(const Intersections& other)
{
	mbrIntersections.clear();
	for (int i = 0; i < other.mbrIntersections.size(); i++)
	{
		setIntersection(other.mbrIntersections[i].get());
		// Intersection varIs = other.mbrIntersections[i];
		// mbrIntersections.push_back(varIs);
	}
	//this->mbrIntersections = other.mbrIntersections;
}
Intersections::Intersections(double t, Form *s)
{
	setIntersection(t, s);
	// Intersection i = Intersection(t,std::make_unique<Form>(*s.get()));
	// mbrIntersections.push_back(i);
}
// Intersections::Intersections(double t, std::unique_ptr<Form> s)
// {
// 	setIntersection(t, s.get());
// 	// Intersection i = Intersection(t,std::make_unique<Form>(*s.get()));
// 	// mbrIntersections.push_back(i);
// }
Intersections& Intersections::operator=(const Intersections other)
{
	if (this == &other) return *this;
	mbrIntersections.clear();
	for (int i = 0; i < other.mbrIntersections.size(); i++)
	{
		setIntersection(other.mbrIntersections[i].get());
		// Intersection varIs = other.mbrIntersections[i];
		// mbrIntersections.push_back(varIs);
	}
	// this->mbrIntersections = other.mbrIntersections;
	return *this;
}
// void Intersections::intersect(double t, std::unique_ptr<Form> argObject)
// {
// 	auto funcComp = [&](const std::unique_ptr<Intersection> &argA, const std::unique_ptr<Intersection> &argB) -> bool {
// 		return argA->mbrTime < argB->mbrTime;
// 	};
// 	setIntersection(t, argObject.get());
// 	//mbrIntersections.push_back(Intersection(t, std::make_unique<Form>(*argObject)));
//     sort(mbrIntersections.begin(), mbrIntersections.end(), funcComp);
// }
void Intersections::intersect(double t, Form* argObject)
{
	auto funcComp = [&](const std::unique_ptr<Intersection> &argA, const std::unique_ptr<Intersection> &argB) -> bool {
		return argA->mbrTime < argB->mbrTime;
	};
	setIntersection(t, argObject);
	//mbrIntersections.push_back(Intersection(t, std::make_unique<Form>(*argObject)));
    sort(mbrIntersections.begin(), mbrIntersections.end(), funcComp);
}
Intersection Intersections::hit()
{
	int index = -1;
	for (int i = 0; i < mbrIntersections.size(); i++)
	{
		if ((mbrIntersections[i]->mbrTime > 0)
			&& ((index < 0) || (mbrIntersections[i]->mbrTime <= mbrIntersections[index]->mbrTime)))
		{
			index = i;
			break;
		}
	}
	if (index < 0) {
		//std::cout << "hit missed" << std::endl;
		Intersection varIx = Intersection();
		varIx.mbrExists = false;
		return varIx;
	}
	else {
		//std::cout << "hit" << std::endl;
		return *mbrIntersections[index].get();
	}
}
void Intersections::setIntersection(Intersection* argIx) {
	mbrIntersections.push_back(std::make_unique<Intersection>(*argIx));
}
void Intersections::setIntersection(double t, Form* argObject) {
	mbrIntersections.push_back(std::make_unique<Intersection>(t, argObject));
}

File: src/Intersection.cs
using LibForm;
using LibTuple;
using LibRay;
using LibProjectMeta;
namespace LibIntersection;

public class IntersectionState {
	public double _fieldTime;
	public Form _fieldObject;
	public Point _fieldPoint;
	public Point _fieldOverPoint;
	public Point _fieldUnderPoint;
	public Vector _fieldPov;
	public Vector _fieldNormal;
	public Vector _fieldReflect;
	public bool _fieldInside;
	public double _fieldRefractiveIndexOne;
	public double _fieldRefractiveIndexTwo;
	public IntersectionState(){
		_fieldTime = 0;
		_fieldObject = new Form();
		_fieldPoint = new Point();
		_fieldPov = new Vector();
		_fieldNormal = new Vector();
		_fieldInside = false;
		_fieldUnderPoint = new Point();
		_fieldReflect = new Vector();
		_fieldOverPoint = new Point();
		_fieldRefractiveIndexOne = 0.0;
		_fieldRefractiveIndexTwo = 0.0;
	}
	public IntersectionState(IntersectionState argOther) {
		_fieldTime = argOther._fieldTime;
		SetObject(argOther._fieldObject);
		_fieldPoint = argOther._fieldPoint;
		_fieldPov = argOther._fieldPov;
		_fieldNormal = argOther._fieldNormal;
		_fieldInside = argOther._fieldInside;
		_fieldOverPoint = argOther._fieldOverPoint;
		_fieldReflect = argOther._fieldReflect;
		_fieldOverPoint = argOther._fieldOverPoint;
		_fieldRefractiveIndexOne = argOther._fieldRefractiveIndexOne;
		_fieldRefractiveIndexTwo = argOther._fieldRefractiveIndexTwo;
	}
	public double GetSchlick() {
		double varEyeNormal = this._fieldPov.GetDotProduct(this._fieldNormal);
		if (this._fieldRefractiveIndexOne > this._fieldRefractiveIndexTwo) {
			double varNToN = this._fieldRefractiveIndexOne/this._fieldRefractiveIndexTwo;
			double varSinTSqr = (varNToN*varNToN) * (1.0-(varEyeNormal*varEyeNormal));
			if (varSinTSqr > 1.0) { return 1.0;}
			varEyeNormal = Math.Sqrt(1.0-varSinTSqr);
		}
		double varR0 = Math.Pow((this._fieldRefractiveIndexOne - this._fieldRefractiveIndexTwo)/(this._fieldRefractiveIndexOne+this._fieldRefractiveIndexTwo),2);
		return varR0 + (1-varR0) * Math.Pow((1-varEyeNormal),5);
	}
    public void SetObject(Form argObject) {
		_fieldObject = argObject;
	}
	public void RenderConsole() {
		Console.WriteLine("IntersectionState::renderConsole()");
		Console.WriteLine($"IntersectionState::mbrTime: {_fieldTime}");
		Console.WriteLine("IntersectionState::mbrObject: ");
		_fieldObject.RenderConsole();
		Console.WriteLine("IntersectionState::mbrPoint: ");
		_fieldPoint.RenderConsole();
		Console.WriteLine("IntersectionState::mbrEye: ");
		_fieldPov.RenderConsole();
		Console.WriteLine("IntersectionState::mbrNormal: ");
		_fieldNormal.RenderConsole();
		Console.WriteLine($"IntersectionState::mbrInside: {_fieldInside}");
		Console.WriteLine("IntersectionState::mbrOverPoint: ");
		_fieldOverPoint.RenderConsole();
		Console.WriteLine("IntersectionState::mbrReflect: ");
		_fieldReflect.RenderConsole();
	}
};

public class Intersection {
	public bool _fieldExists;
	public double _fieldTime;
	public Form _fieldObject;
	protected ProjectMeta varPM = new ProjectMeta();
	public Intersection()
	{
		_fieldObject = new Form();
		_fieldTime = 0;
		_fieldExists = true;
	}
	public Intersection(Intersection argOther){
		SetObject(argOther._fieldObject);
		_fieldTime = argOther._fieldTime;
		_fieldExists = argOther._fieldExists;
	}
	public Intersection(double argTime, Form argObj)
	{
		SetObject(argObj);
		_fieldTime = argTime;
		_fieldExists = true;
	}
	public bool CheckEqual(Intersection argOther)
	{
		return _fieldTime == argOther._fieldTime && _fieldObject.CheckEqual(argOther._fieldObject) && _fieldExists == argOther._fieldExists;
	}
	public IntersectionState GetState(Ray argRay, List<Intersection> argIntersections)
	{
		IntersectionState varIs = new IntersectionState();
		varIs._fieldTime = _fieldTime;
		varIs.SetObject(_fieldObject);
		varIs._fieldPoint = argRay.GetPosition(varIs._fieldTime);
		varIs._fieldPov = -argRay._fieldDirection;
		varIs._fieldNormal = varIs._fieldObject.GetNormal(varIs._fieldPoint);
		if (varIs._fieldNormal.GetDotProduct(varIs._fieldPov) < 0) {
			varIs._fieldInside = true;
			varIs._fieldNormal = -varIs._fieldNormal;
		}
		else {
			varIs._fieldInside = false;
		}
		varIs._fieldOverPoint = varIs._fieldPoint + (varIs._fieldNormal * varPM.getEpsilon());
		varIs._fieldUnderPoint = varIs._fieldPoint - (varIs._fieldNormal * varPM.getEpsilon());
		varIs._fieldReflect = argRay._fieldDirection.GetReflect(varIs._fieldNormal);
		List<Form> varHitObjects = new List<Form>();
		Intersection varHit = new Intersection(varIs._fieldTime, varIs._fieldObject);
		for (int i = 0; i < argIntersections.Count; ++i) {
			if (varHit.CheckEqual(argIntersections[i])) {
				if (varHitObjects.Count == 0) { varIs._fieldRefractiveIndexOne = 1.0; }
				else { varIs._fieldRefractiveIndexOne = varHitObjects[varHitObjects.Count-1]._fieldMaterial._fieldRefractiveIndex; }
			}
			int varFound = -1;
			for (int j = 0; j < varHitObjects.Count; ++j)
			{
				if (varHitObjects[j].CheckEqual(argIntersections[i]._fieldObject)) {
					varFound = j;
					break;
				}
			}
			if (varFound < 0) { varHitObjects.Add(argIntersections[i]._fieldObject); }
			else { varHitObjects.RemoveAt(varFound); }
			if (varHit.CheckEqual(argIntersections[i])) {
				if (varHitObjects.Count == 0) { varIs._fieldRefractiveIndexTwo = 1.0; }
				else { varIs._fieldRefractiveIndexTwo = varHitObjects[varHitObjects.Count-1]._fieldMaterial._fieldRefractiveIndex; }
				break;
			}
		}
		return varIs;
	}
	public void SetObject(Form argObject) {
		_fieldObject = argObject;
	}
}

public class Intersections {
    public List<Intersection> _fieldIntersections;
	public Intersections()
	{
		_fieldIntersections = new List<Intersection>();
	}
	public Intersections(Intersections argOther)
	{
		_fieldIntersections = new List<Intersection>();
		for (int i = 0; i < argOther._fieldIntersections.Count(); ++i)
		{
			SetIntersection(argOther._fieldIntersections[i]);
		}
	}
	public Intersections(double argTime, Form argObj)
	{
		_fieldIntersections = new List<Intersection>();
		SetIntersection(argTime, argObj);
	}
	public void SetIntersect(double time, Form argObject)
	{
		SetIntersection(time, argObject);
		_fieldIntersections.Sort(delegate(Intersection argA, Intersection argB) {
			return argA._fieldTime.CompareTo(argB._fieldTime);
		});
	}
	public Intersection GetHit() {
		int index = -1;
		for (int i = 0; i < _fieldIntersections.Count(); ++i)
		{
			if ((_fieldIntersections[i]._fieldTime > 0)
				&& ((index < 0) || (_fieldIntersections[i]._fieldTime <= _fieldIntersections[index]._fieldTime)))
			{
				index = i;
				break;
			}
		}
		if (index < 0) {
			Intersection varIx = new Intersection();
			varIx._fieldExists = false;
			return varIx;
		}
		else {
			return _fieldIntersections[index];
		}
	}
	protected void SetIntersection(Intersection argIx) {
		_fieldIntersections.Add(argIx);
	}
    protected void SetIntersection(double argTime, Form argObject) {
		_fieldIntersections.Add(new Intersection(argTime, argObject));
	}
};